src/cargo/util/errors.rs

```diff
@@ -286,6 +286,20 @@ impl From<Box<CargoError>> for CliError {
 }
 
 // =============================================================================
+// NetworkError trait
+
+pub trait NetworkError: CargoError {
+    fn maybe_spurious(&self) -> bool;
+}
+
+impl NetworkError for git2::Error {
+    fn maybe_spurious(&self) -> bool { true }
+}
+impl NetworkError for curl::ErrCode {
+    fn maybe_spurious(&self) -> bool { true }
```

Comment:
Ah using the raw error codes from curl is fine, this should basically just be a whitelist of known bad error codes that can be retried, and we can expand the list over time.